sdk/rust/core/src/lib.rs

```diff
@@ -1,34 +1,71 @@
+//! - Parsers for VAA's and Payloads.
+//! - Data types for Wormhole primitives such as GuardianSets and signatures.
+//! - Verification Primitives for securely checking payloads.
+
+#![deny(warnings)]
 #![deny(unused_results)]
 
-pub use chain::*;
-pub use error::*;
-pub use vaa::*;
+use serde::{Deserialize, Serialize};
 
-pub mod chain;
+mod arraystring;
+mod chain;
+pub mod core;
+pub mod nft;
+mod serde_array;
+pub mod token;
 pub mod vaa;
 
-#[macro_use]
-pub mod error;
-
-/// Helper method that attempts to parse and truncate UTF-8 from a byte stream. This is useful when
-/// the wire data is expected to contain UTF-8 that is either already truncated, or needs to be,
-/// while still maintaining the ability to render.
-///
-/// This should be used to parse any Text-over-Wormhole fields that are meant to be human readable.
-pub(crate) fn parse_fixed_utf8<T: AsRef<[u8]>, const N: usize>(s: T) -> Option<String> {
-    use bstr::ByteSlice;
-    use std::io::Cursor;
-    use std::io::Read;
-
-    // Read Bytes.
-    let mut cursor = Cursor::new(s.as_ref());
-    let mut buffer = vec![0u8; N];
-    cursor.read_exact(&mut buffer).ok()?;
-    buffer.retain(|&c| c != 0);
-
-    // Attempt UTF-8 Decoding. Stripping invalid Unicode characters (0xFFFD).
-    let mut buffer: Vec<char> = buffer.chars().collect();
-    buffer.retain(|&c| c != '\u{FFFD}');
-
-    Some(buffer.iter().collect())
+pub use {chain::Chain, vaa::Vaa};
+
+/// The `GOVERNANCE_EMITTER` is a special address Wormhole guardians trust to observe governance
+/// actions from. The value is "0000000000000000000000000000000000000000000000000000000000000004".
+pub const GOVERNANCE_EMITTER: Address = Address([
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04,
+]);
+
+#[derive(
+    Serialize, Deserialize, Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash,
+)]
+pub struct GuardianAddress(pub [u8; 20]);
+
+/// Wormhole specifies addresses as 32 bytes. Addresses that are shorter, for example 20 byte
+/// Ethereum addresses, are left zero padded to 32.
+#[derive(
+    Serialize, Deserialize, Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash,
+)]
+pub struct Address(pub [u8; 32]);
+
+/// Wormhole specifies an amount as a uint256 encoded in big-endian order.
+#[derive(
+    Serialize, Deserialize, Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash,
+)]
+pub struct Amount(pub [u8; 32]);
+
+/// A `GuardianSet` is a versioned set of keys that can sign Wormhole messages.
+#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct GuardianSetInfo {
+    /// The set of guardians public keys, in Ethereum's compressed format.
+    pub addresses: Vec<GuardianAddress>,
+
+    /// How long after a GuardianSet change before this set is expired.
+    #[serde(skip)]
```

Comment:
Why is this being skipped? This is still required when serializing the account. We should drop this if it's not meant to be used.

Reply:
It's skipped because it doesn't appear in the actual GuardianSetUpgrade governance messages but is still used by contracts when persisting.  I'll remove it in a follow-up PR.